### DISCUSSION TITLE
Trying to understand how Renovate detects versions in a Dockerfile

### DISCUSSION THREAD

**Discussion by p3k** [renovatebot/renovate#23478] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

35.159.4

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

It works, and I do not know why – but would like to understand.

### Wanted end result.

We have a Dockerfile defining dependencies like so:

```Dockerfile
ARG DEBIAN_VERSION=12.0
ARG DOCKER_VERSION=24.0.2
ARG JAVA_VERSION=17.0.7_7
ARG NODE_VERSION=18.16.1

FROM debian:${DEBIAN_VERSION}-slim
FROM docker:${DOCKER_VERSION} AS docker
FROM eclipse-temurin:${JAVA_VERSION}-jdk AS java
FROM node:${NODE_VERSION}-bookworm-slim AS nodejs
…
```

I know about the regex manager that can be used for resolving such cases using comments like `# renovate: datasource=… depName=…`.

However, so far I have not used such a regex manager – nevertheless, Renovate is already detecting the dependencies and versions just fine!

So, how is Renovate doing this?

### What you tried so far.

I searched through the repo and the docs, but so far did not find any hint. Maybe it’s a simple question, and I am just looking at all the wrong places…?

### Relevant debug logs

<details><summary>Logs</summary>

```json
{
	"dockerfile": [
		{
			"deps": [
				{
					"depName": "docker",
					"currentValue": "24.0.2",
					"replaceString": "ARG DOCKER_VERSION=24.0.2\n",
					"autoReplaceStringTemplate": "ARG DOCKER_VERSION={{#if newValue}}{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}\n",
					"datasource": "docker",
					"depType": "stage",
					"updates": [
						{
							"bucket": "non-major",
							"newVersion": "24.0.4",
							"newValue": "24.0.4",
							"newMajor": 24,
							"newMinor": 0,
							"updateType": "patch",
							"branchName": "renovate/docker-24.x"
						}
					],
					"packageName": "docker",
					"versioning": "docker",
					"warnings": [],
					"registryUrl": "https://index.docker.io",
					"currentVersion": "24.0.2",
					"isSingleVersion": true,
					"fixedVersion": "24.0.2"
				},
				{
					"depName": "eclipse-temurin",
					"currentValue": "17.0.7_7-jdk",
					"replaceString": "eclipse-temurin:17.0.7_7-jdk",
					"autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
					"datasource": "docker",
					"depType": "stage",
					"updates": [],
					"packageName": "eclipse-temurin",
					"versioning": "regex:^(?<major>\\d+)?(\\.(?<minor>\\d+))?(\\.(?<patch>\\d+))?([\\._+](?<build>\\d+))?(-(?<compatibility>.*))?$",
					"warnings": [],
					"registryUrl": "https://index.docker.io",
					"currentVersion": "17.0.7_7-jdk",
					"fixedVersion": "17.0.7_7-jdk"
				},
				{
					"depName": "node",
					"currentValue": "18.16.1-bookworm-slim",
					"replaceString": "node:18.16.1-bookworm-slim",
					"autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
					"datasource": "docker",
					"depType": "stage",
					"updates": [
						{
							"bucket": "non-major",
							"newVersion": "18.17.0",
							"newValue": "18.17.0-bookworm-slim",
							"newMajor": 18,
							"newMinor": 17,
							"updateType": "minor",
							"branchName": "renovate/node-18.x"
						},
						{
							"bucket": "major",
							"newVersion": "20.4.0",
							"newValue": "20.4.0-bookworm-slim",
							"newMajor": 20,
							"newMinor": 4,
							"updateType": "major",
							"branchName": "renovate/node-20.x"
						}
					],
					"packageName": "node",
					"versioning": "docker",
					"warnings": [],
					"sourceUrl": "https://github.com/nodejs/node",
					"registryUrl": "https://index.docker.io",
					"currentVersion": "18.16.1",
					"isSingleVersion": true,
					"fixedVersion": "18.16.1-bookworm-slim"
				},
				{
					"depName": "debian",
					"currentValue": "12.0-slim",
					"replaceString": "debian:12.0-slim",
					"autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
					"datasource": "docker",
					"depType": "final",
					"updates": [],
					"packageName": "debian",
					"versioning": "docker",
					"warnings": [],
					"registryUrl": "https://index.docker.io",
					"currentVersion": "12.0",
					"fixedVersion": "12.0-slim"
				}
			],
			"packageFile": "tools/docker-development/Dockerfile"
		}
	]
}
```

</details>

**Comment by p3k** [comment]
Oh, forgot: maybe an excerpt from the Renovate config is a good idea.

```json
{
	"$schema": "https://docs.renovatebot.com/renovate-schema.json",

	"extends": [
		"config:base",
		"github>whitesource/merge-confidence:beta",
		"npm:unpublishSafe",
		"schedule:weekdays",
		":disableRateLimiting",
		":pinVersions",
		":semanticCommitsDisabled"
	],

	"dryRun": null,
	"enabledManagers": ["dockerfile", "gradle", "gradle-wrapper", "npm"]
}
```

**Comment by Churro** [comment, marked as answer]
The behavior you describe is supported natively (no preset required) and was introduced with PR https://github.com/renovatebot/renovate/pull/15751. Its current implementation can be found [here](https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/dockerfile/extract.ts).

The identification of images with variables works basically like this:

- Process a `Dockerfile` line-wise and collect all `ARG` definitions
- Whenever a `FROM` statement with variables is found, renovate attempts to resolve them using the `ARG` values it has seen so far
- If successful, renovate marks the lines as contributing to a container image reference
- If all variables could be resolved, the full image definition is split into chunks again to find out if the now known version and digest are located in variables. The intention here is to find only the few lines that may change in case of image updates.
- Based on this output, the auto-replace functionality of renovate is provided with a templated Dockerfile fragment that describes exactly where to replace version and digest.

**Reply by p3k** [reply]
Thanks a lot for the explanation and insights, @Churro!

Is this documented somewhere except for the PR?

**Reply by Churro** [reply]
> Is this documented somewhere except for the PR?

I don't think so. Internals how renovate operates are barely documented, probably also because this may change at any point with new PRs.

**Reply by p3k** [reply]
It makes sense to at least mention this feature because without its knowledge and by looking at the documentation of [`regexManagers:dockerfileVersions`](https://docs.renovatebot.com/presets-regexManagers/#regexmanagersdockerfileversions) one could get the impression the regex manager is the reason for finding the Docker versions.
